### ISSUE TITLE
Allow Map IDs and Map Names to be functionally interchangeable.

### ISSUE THREAD

**Issue by FullBleed** [RPTools/maptool#2647]
_**Is your feature request related to a problem? Please describe.**_

Map names, like token names, can sometimes give away important information before players should have it.  Given that, I often want to use misdirection or bland naming schemes ("Temple Ruins") until players know more about their actual locale ("The Forgotten Temple of Tharizdun").  But, changing the names can cause problems with a variety of potential features that may currently expect a map's name to be a static identifier (since the map's name is the only way to identify the map functionally right now.)   

_**Describe the solution you'd like**_

I would like functions that currently take a map name to also work with the map's existing ID.  That way people can dynamically change a map's name without compromising anything that needs a static map identifier.  I'm not sure how some token functions can take both a token name and/or ID, but I'm hoping that the lookup methodology might already exist and not require touching every function that currently takes a map name. 

_**Describe alternatives you've considered**_

Give maps some kind of alternative "Display Name"?

**Comment by michaelperino** [comment]
Where is the map name being displayed that you consider a spoiler? Only in the title of the program at the top, or also in the 'Select Map' drop-down?

**Comment by FullBleed** [comment]
> Where is the map name being displayed that you consider a spoiler? Only in the title of the program at the top, or also in the 'Select Map' drop-down?

Any place a player can see the name.  So, potentially, both cases.

**Comment by michaelperino** [comment]
Discussed my possible approach with FullBleed, my bad for not commenting earlier. Basically looked at the alternative idea he mentioned as a solution. A row could be added to map property editor that allows the display name to be edited differently.

![image](https://user-images.githubusercontent.com/33206261/118388531-bcd54400-b5ea-11eb-8b0d-d21ad9fa3f1c.png)

The players on the server will then see the labels on their map selection and top of the screen compared to the GM seeing the real name. Note that the text Player Map Alias was updated as shown above to hopefully make it more clear to GMs what is being accomplished. If this is a liked change, let me know and I'll think of a better way to say it again because "Player Display Name" could reference a couple different things.

![image](https://cdn.discordapp.com/attachments/843353001179480084/843361168277372928/unknown.png)

**Comment by FullBleed** [comment]
Been thinking about this a bit and we might be able to get away with just calling the new field "Display Name".

Some other thoughts:

1)  Should definitely have a tooltip there to let people know what the Display Name is.
2)  On the GM side, when you have a Display Name set I think it should be first in the application title bar with the real name in italicized parenthesis:  Temple _(Temple of Death)_.  And probably the same in the Select Map drop-down.  GMs often have to tell players to "switch to X" and having the name the Players are seeing as "primary" will help keep that simple.  And putting the real/GM name in parenthesis is consistent with how the GM Name in tokens is displayed. 
3)  Will need get/set functions for the new field.  Not sure if that would be best handled in a separate issue.

Thanks for tackling this.

**Comment by michaelperino** [comment]
All excellent points! Here's what things are currently looking like:
![image](https://user-images.githubusercontent.com/33206261/118392198-f3688a00-b5fd-11eb-962e-6698869bc568.png)

1. A tooltip has been included.
2. I agree about showing both. You will also notice that on Cracked Mud where there was no needed display name, no parentheses are included. However, I'm unsure about the order of the Name and Display Name. I have left sorting to use the Name because it allows the GM to group maps together or apply a custom order. I'm not sure how to make that obvious though. I'm also not sure whether to do sorting differently for Players or if their sorting should be the same as the GM's (if for no other reason than keeping the DM's that stream and know their map order sane).
3. By functions I'm thinking you mean macro functions? The Zone class has functions to set that but I could look into macro functions.

I think there is a small bug. When I edit a map as a GM on a server, it moves the players to the top map on the list. I'll try to track that down tomorrow if the community wants the feature.

**Comment by Phergus** [comment]
Replaces/overlaps with issues #2044 and #2033

**Comment by michaelperino** [comment]
Thanks for linking back to that discussion, brings up an excellent point that I should invert Name / Display Name positions to match the order on tokens. If we want to look at a multi line setup for the selector that's probably possible, but I'm not sure it's necessary on even a 1080 monitor (no clue how many play on 768 or 720).

I'll track down the bug mentioned above when I wake up, feature should work pretty much as intended for now.

Any idea why Azhrei thought importing maps with an extra string saved wouldn't work? They mentioned it in 2044, but I was able to successfully import the map exported from dev into 1.8.5, import into dev -> dev without losing data, and import 1.8.5 -> dev as intended.

**Comment by FullBleed** [comment]
Mike, for the record, when I mention "functions" I'm almost always going to be referring to macro functions.  That's where most user's rubber meets the road.


On the naming conventions...

This is in response to seeing the "Map Name" showing as "GM Name":

It's unfortunate that tokens and maps didn't start with both using Name, GM Name, and IDs equally.  But they don't.

One of the concerns I had with changing the current Map Name to "GM Name" is that getting it to jive with token nomenclature isn't really possible.  For example: 

Token GM Name:   Only ever seen by the GM.  Functionally used with get/set tokenGMName().

But currently the Map Name is used with *every* function that refers to a map (which is a *lot* of functions.)  It's more similar to the Token Name functionally (though most token functions use the ID and only some use the name, too.)

But a new Map Display Name will, functionaly, only be used with a get/set mapDisplayName(). [So more like the Token GM Name functionally.]

So I think it's better to leave "Map Name" alone since nothing is changing there.

**Comment by Azhrei** [comment]
> Any idea why Azhrei thought importing maps with an extra string saved wouldn't work?
> 
Because I hadn’t tried it and couldn’t remember how XStrram handled fields in the XML that didn’t exist in the Java object being deserialized.

But if it works, then all’s good. 👍
--
Frank Edwards
Edwards & Edwards Consulting, LLC
Hours:  10A-6P ET
Phone:  (813) 406-0604

**Comment by FullBleed** [comment]
I still think that the Display name should precede the Map Name (with the real map name being in parenthesis.)  The Display Name is the "common" name between player/gm and how the maps should be referred to by both until the Map Name becomes the Display Name.  ;)

I suspect that there will be GMs that want the list to be sorted by Display Name or Name... so there probably should to be a toggle somewhere.

Also, I don't think we *need* a multi-line display in the map drop down... at least, not until someone indicates they have a problem with it.  It might be moderately more aesthetically pleasing... but it wil also stretch out the map list and require more scrolling (which I know I don't want.)

**Comment by michaelperino** [comment]
> But currently the Map Name is used with every function that refers to a map (which is a lot of functions.) It's more similar to the Token Name functionally (though most token functions use the ID and only some use the name, too.)
But a new Map Display Name will, functionaly, only be used with a get/set mapDisplayName(). [So more like the Token GM Name functionally.]

Do you think it should be opposite the token then? With the real name (traditional/GM name) on top and "Display Name" on bottom?

> I still think that the Display name should precede the Map Name (with the real map name being in parenthesis.) The Display Name is the "common" name between player/gm and how the maps should be referred to by both until the Map Name becomes the Display Name. ;)

Not a bad way to think about it.

> I suspect that there will be GMs that want the list to be sorted by Display Name or Name... so there probably should to be a toggle somewhere.

I'll look into a setting.

> Also, I don't think we need a multi-line display in the map drop down... at least, not until someone indicates they have a problem with it. It might be moderately more aesthetically pleasing... but it wil also stretch out the map list and require more scrolling (which I know I don't want.)

100% agree

**Comment by FullBleed** [comment]
> > But currently the Map Name is used with every function that refers to a map (which is a lot of functions.) It's more similar to the Token Name functionally (though most token functions use the ID and only some use the name, too.)
> > But a new Map Display Name will, functionaly, only be used with a get/set mapDisplayName(). [So more like the Token GM Name functionally.]
> 
> Do you think it should be opposite the token then? With the real name (traditional/GM name) on top and "Display Name" on bottom?

If I understand the question, I think the way you have it now with Map Name first and the Display Name being the next field is the way to go.

**Comment by FullBleed** [comment]
Quick question, how does a commit here end up "In Progress" in a current project (in this case 1.9)?

**Comment by Phergus** [comment]
@FullBleed Commits to a contributors fork don't necessarily trigger an issue being moved to In Progress.  Once the proposed changes/functionality have been reviewed by a core dev and approved to be included in a particular build the issue will be assigned to the contributor and marked as In Progress at that point.

If I'm following what has been said so far:

- (Map) Name will stay the primary name for a map and will be the name used in macro functions as it is now.
- A new `Display Name` field will be added to the Map Properties dialog with appropriate localization.
- The `Display Name` field will now be the primary display name in the Title Bar and Map Selector.
    - In the Title Bar it will be shown as `Display Name (Map Name)` for GM clients.
    - In Map Selector, Player clients will only see `Display Name` while GM clients get the `Display Name (Map Name)`.
 - Macro functions for `getMapDisplayName()` / `setMapDisplayName()` will be provided.
 - `getInfo("map")` will be updated to include `Display Name` in returned JSON object.
 - Wiki documentation will be created/updated for the above.

Does that sound good @michaelperino ?

**Comment by michaelperino** [comment]
I believe that is correct. Once I wrap that copyToken thing I'll go through and document exactly how the changes look to avoid any confusion and to get feedback easier. Remind me either here or on discord who I need to talk to about getting a wiki account.

**Comment by Phergus** [comment]
@melek Can you coordinate with Michael about getting him a wiki account?

**Comment by michaelperino** [comment]
Here's the changes I can think of that I've made. Let me know if anything looks weird.
[2647 work completed.pdf](https://github.com/RPTools/maptool/files/6565777/2647.work.completed.pdf)

(last paragraph copied from document) Quick question regarding work on maptool in general: is there a process needed to prepare for localization? Currently I’ve created new i18n fields for my changes and included them in the English i18n file as seen below, but should I also create those fields in the files of other languages so translators know that they are there and need to be changed? 
![image](https://user-images.githubusercontent.com/33206261/120093734-badaad00-c0e1-11eb-92e1-5efecde73df4.png)

**Comment by cwisniew** [comment]
@michaelperino no need to create in the other files. This file will be read by crowdin (site used for translation) and inform those that have subscribed to such notifications that there are new strings to be translated.

**Comment by Phergus** [comment]
Thanks for the write-up!

1. `getPlayerAlias()` and `setPlayerAlias()`.  Are those methods in Zone?
2. Sounds good.
3. I'm not sure about using "True Name" but can't come up with a better term so 
4. One thought is that instead `getCurrentMapDisplayName()` just have `getMapDisplayName()` which if no argument is provided returns the Display Name for the current map or an argument is provided returns the display name for that map.
5. Sounds good.
6. As Craig mentioned above.  Just the base i18n file needs to be updated.

**Comment by michaelperino** [comment]
1. Yes
4. I like that, I'll look into it.

I'm having an error with the setMapDisplayName as a server function which I'll take a closer look at when I have time tomorrow.

**Comment by FullBleed** [comment]
> 
> 
> Here's the changes I can think of that I've made. Let me know if anything looks weird.

Love it.  You have a thorough and detailed work flow process and don't seem afraid to put your hands in some areas many devs like to avoid (like ui/settings)... really happy to see you contributing to MT, Mike.

Here's to seeing even more from you!

**Comment by michaelperino** [comment]
I'm so close btw. Only things left are documentation and one error. I'm getting a weird error where the server method's handler's parameters are not working well with my function. With the normal setMapName() Function, the first parameter of ```parameters``` is an object with the caller of the function. In mine, it is being passed a GUID in place of the caller which is not converting to string (rightfully so as that's not what we want anyway). Hopefully I can get this knocked out within a couple days as work is picking up.

**Comment by michaelperino** [comment]
This should now be done besides documentation with that bug fixed. I'll work on that tonight.

**Comment by Phergus** [comment]
Sounds good.  I'll keep an eye out for a PR.

**Comment by michaelperino** [comment]
Documentation should be in a good state. https://wiki.rptools.info/index.php/Special:Contributions/MichaelP

**Comment by Phergus** [comment]
Testing.

- Load/Save Campaign good.
- Load previous Campaigns good.
- Load 1.9.0 campaign into 1.8.5 no errors. Save from 1.8.5 and reload into 1.9.0 no errors and display names lost as expected.
- Export/Import of 1.9.0 map good.  Import of 1.9.0 map into 1.8.5 no errors and vice-versa.
- `getInfo()` output good.
- New macro functions working as expected except for `getAllMapDisplayNames()`.
- No issues seen with existing macro functions for maps.

**Potential Issue**
`getAllMapDisplayNames()` returns all map names, even hidden ones, when run by a player via a Campaign macro.  While it doesn't seem likely that a GM would put such a macro on the Campaign panel.  A player could also run the macro from a Token macro if they set it up beforehand on a token.  Attempting to run it from a Global macro correctly stops the function from being run.

**Comment by michaelperino** [comment]
Interesting, I thought that would be the desired behavior for any macros the GM creates that rely on all the maps. Got any macros off the top of your head that would behave like you describe (not allowing players to activate even from non-editable macros)?

**Comment by Phergus** [comment]
Off the top of my head, no.  And checking `getAllMapNames()` and `getTokenNames()` they both will return the names of maps and tokens that are hidden from a player when run from a trusted macro on the Campaign panel.  So this isn't really a new issue but is a potential problem for a GM who unwittingly uses those in a campaign macro without filtering the output. Or an alternate scenario where a player creates a token with those macros already on it and drops it into a campaign.  It's not uncommon for GMs to have the players manage their own tokens.

In any case, this isn't really a new issue and is certainly an edge case that can also be seen with any macro function that can reveal information that is normally hidden from the players.

**Comment by michaelperino** [comment]
Cool, if you'd like me to look into all those ```getAll___()``` functions, that's probably worth a different issue. I'm in between issues if you want to assign it to me.

**Comment by Phergus** [comment]
I'll think about how to frame the issue for discussion and possible resolution.

**Comment by Phergus** [comment]
Documentation completed.

**Comment by FullBleed** [comment]
>Or an alternate scenario where a player creates a token with those macros already on it and drops it into a campaign. It's not uncommon for GMs to have the players manage their own tokens.
> 
> In any case, this isn't really a new issue and is certainly an edge case that can also be seen with any macro function that can reveal information that is normally hidden from the players.

I can think of one solution for this that is a feature I'd like to see but don't think there is a FREQ for: The ability to deny player's the right to drop assets on a map.  The potential for abuse is high and you never know what unauthorized tools could be introduced into your game.  This is just another example of a potential exploitation (and I'm sure there are more things the informed macro writer could sneak in.)  I prefer to have complete control over the table and think we should have the option.

**Comment by Phergus** [comment]
Tested.
